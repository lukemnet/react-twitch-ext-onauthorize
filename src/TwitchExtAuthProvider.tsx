import React, { ReactElement, useState, useEffect } from 'react';

export interface TwitchAuthResponse {
  channelId: string;
  clientId: string;
  token: string;
  userId: string;
}

export interface TwitchAuthObject extends TwitchAuthResponse {
  authorized: boolean;
}

export interface TwitchAuthProviderProps {
  unauthorized?: ReactElement,
  enable?: boolean,
  children: (auth: TwitchAuthObject) => ReactElement
}

interface TwitchContext extends Window {
  [key: string]: any;
  Twitch: {
    [key:string]: any;
    ext: {
      [key:string]: any;
      onAuthorized: (callback: CallableFunction) => void;
    }
  }
}

const TwitchAuthProvider = (props: TwitchAuthProviderProps) => {
  const enable = (typeof props.enable !== 'undefined') ? props.enable : true;
  const unauthorized = props.unauthorized || <div>Unauthorized</div>;
  const [auth, setAuth] = useState({ authorized: false, channelId: '', token: '', clientId: '', userId: '' });

  useEffect(() => {
    const getTwitchExtensionContext = (callback: CallableFunction) => {
      if (enable) {
        const twitchContext = window as TwitchContext;
        const itsTwitch = !!(twitchContext.Twitch && twitchContext.Twitch.ext);
        if (itsTwitch) {
          twitchContext.Twitch.ext.onAuthorized((twitchAuth: TwitchAuthResponse) => {
            callback({ authorized: true, ...twitchAuth })
          });
        }
      } else {
        callback({
          authorized: true,
          channelId: 'disabled',
          clientId: 'disabled',
          token: 'disabled',
          userId: 'disabled',
        });
      }
    }

    getTwitchExtensionContext((twitchAuthObject: TwitchAuthObject) => {
      setAuth(twitchAuthObject);
    });
  }, [enable]);
  
  return auth.authorized
    ? props.children(auth)
    : unauthorized;
}

export default TwitchAuthProvider;