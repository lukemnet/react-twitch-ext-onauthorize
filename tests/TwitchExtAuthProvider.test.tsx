import React from 'react';
import { mount } from 'enzyme';
import TwitchExtAuthProvider, { TwitchAuthObject } from '../src';

const TestElement = (twitchAuth: TwitchAuthObject) => <div>TestElement {JSON.stringify(twitchAuth)}</div>;
const UnauthorizedTestElement = () => <div>UnauthorizedTestElement</div>;

const mockTwitchExtensionContext = () => {
  Object.defineProperty(window, 'Twitch', {
    value: {
      ext: {
        onAuthorized: jest.fn().mockImplementation((callback: Function) => {
          callback({
            authorized: true,
            channelId: 'test_channel_id',
            token: 'test_token',
            clientId: 'test_client_id',
            userId: 'test_user_id',
          });
      }),
      },
    },
    writable: true,
  });
}

const resetTwitchExtensionContext = () => {
  (window as any).Twitch = null;
}

describe('TwitchAuthProvider', () => {
  it('should render child element inside a Twitch Extension context when no props are passed', () => {
    mockTwitchExtensionContext();

    const element = mount(
      <TwitchExtAuthProvider>
         {(twitchAuth: TwitchAuthObject) => <TestElement {...twitchAuth} />}
       </TwitchExtAuthProvider>
    );
    expect(element).toMatchSnapshot();
  });

  it('should render child element inside a Twitch Extension context when disabled', () => {
    mockTwitchExtensionContext();

    const element = mount(
      <TwitchExtAuthProvider enable={false}>
         {(twitchAuth: TwitchAuthObject) => <TestElement {...twitchAuth} />}
       </TwitchExtAuthProvider>
    );
    expect(element).toMatchSnapshot();
  });

  it('should render as unauthorized outside of a Twitch Extension context', () => {
    resetTwitchExtensionContext();
    
    const element = mount(
      <TwitchExtAuthProvider>
         {(twitchAuth: TwitchAuthObject) => <TestElement {...twitchAuth} />}
       </TwitchExtAuthProvider>
    );
    expect(element).toMatchSnapshot();
  });

  it('should render custom unauthorized element if passed as a prop outside of a Twitch Extension context', () => {
    resetTwitchExtensionContext();
    
    const element = mount(
      <TwitchExtAuthProvider unauthorized={<UnauthorizedTestElement />}>
         {(twitchAuth: TwitchAuthObject) => <TestElement {...twitchAuth} />}
       </TwitchExtAuthProvider>
    );
    expect(element).toMatchSnapshot();
  });
});