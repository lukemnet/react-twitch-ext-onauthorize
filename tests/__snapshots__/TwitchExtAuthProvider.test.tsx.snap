// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TwitchAuthProvider should render as unauthorized outside of a Twitch Extension context 1`] = `
<TwitchAuthProvider>
  <div>
    Unauthorized
  </div>
</TwitchAuthProvider>
`;

exports[`TwitchAuthProvider should render child element inside a Twitch Extension context when disabled 1`] = `
<TwitchAuthProvider
  enable={false}
>
  <TestElement
    authorized={true}
    channelId="disabled"
    clientId="disabled"
    token="disabled"
    userId="disabled"
  >
    <div>
      TestElement 
      {"authorized":true,"channelId":"disabled","clientId":"disabled","token":"disabled","userId":"disabled"}
    </div>
  </TestElement>
</TwitchAuthProvider>
`;

exports[`TwitchAuthProvider should render child element inside a Twitch Extension context when no props are passed 1`] = `
<TwitchAuthProvider>
  <TestElement
    authorized={true}
    channelId="test_channel_id"
    clientId="test_client_id"
    token="test_token"
    userId="test_user_id"
  >
    <div>
      TestElement 
      {"authorized":true,"channelId":"test_channel_id","token":"test_token","clientId":"test_client_id","userId":"test_user_id"}
    </div>
  </TestElement>
</TwitchAuthProvider>
`;

exports[`TwitchAuthProvider should render custom unauthorized element if passed as a prop outside of a Twitch Extension context 1`] = `
<TwitchAuthProvider
  unauthorized={<UnauthorizedTestElement />}
>
  <UnauthorizedTestElement>
    <div>
      UnauthorizedTestElement
    </div>
  </UnauthorizedTestElement>
</TwitchAuthProvider>
`;
